template.yaml

```diff
@@ -61,12 +79,12 @@ Resources:
                 Effect: Allow
 
 Outputs:
-  # ServerlessRestApi is an implicit API created out of Events key under Serverless::Function
-  # Find out more about other implicit resources you can reference within SAM
-  # https://github.com/awslabs/serverless-application-model/blob/master/docs/internals/generated_resources.rst#api
-  ProgramCodeRulesApi:
+  ApiOrigin:
     Description: "API Gateway endpoint URL for Prod stage for function"
-    Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/program-codes/"
+    Value: !Sub "https://${RestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/"
+  ProgramCodeRulesEndpoint:
+    Description: "Custom endpoint URL for Prod stage for function"
+    Value: !Sub "https://${DnsName}/program-codes/"
```

Comment:
i guess setting up the `RestApi` resource enables this?  

I'm not clear on the idea behind this change.  are we setting this up so that we forgo a redirect to this endpoint or are we setting this up just to get a more friendly URL for the endpoint?

Reply:
Yeah we're creating an explicit RestApi in order to set up a friendly URL - specifically to configure API Gateway to accept TLS connections at our friendly name (we still need to set up a DNS CNAME in org-formation)

Reply:
An alternative to creating an explicit RestApi would be to create a CloudFront Distribution for handling the TLS handshake

Reply:
The cert needs to be manually validated, so we are creating a single wildcard cert in org-formation instead of needing to manually validate one of the resources in this stack.

And yeah, the plan is still to set up a dns forward for this friendly url we are giving the lambda.

Reply:
> we are creating a single wildcard cert in org-formation instead of needing to manually validate one of the resources in this stack

ok, we are setting a cert in org-formation because it's a shared cert.

> the plan is still to set up a dns forward for this friendly url we are giving the lambda

Why do we need to setup a dns forward when we already have a friendly url that's setup here?  Why don't we reference this friendly url directly?

Another approach is to allow AWS to give the url some random url.  We then would setup forwarding from a friendly url  in SageIT to forward to this random url.   I think i prefer this approach so that we don't need to make sure that the url is unique.

Reply:
> Why do we need to setup a dns forward when we already have a friendly url that's setup here? Why don't we reference this friendly url directly?

What we are configuring here is the host name that API Gateway will listen for, and a TLS certificate valid for that host name, but we still need to configure DNS so that client traffic goes to our API Gateway.

> Another approach is to allow AWS to give the url some random url. We then would setup forwarding from a friendly url in SageIT to forward to this random url. I think i prefer this approach so that we don't need to make sure that the url is unique.

Isn't this just using S3 to automate the creation of DNS for us? And isn't it still only automatic when creating that bucket in the SageIT account? What's the benefit of using S3 for an HTTP redirect across account boundaries over creating an ACM certificate in the same account as the API Gateway?

Reply:
ok.  i think i understand now. thanks for the clarifications.